rust/processor/src/gap_detector.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    models::processor_status::ProcessorStatus,
+    processors::ProcessedVersions,
+    schema::processor_status,
+    utils::{
+        database::{execute_with_better_error, PgDbPool},
+        util::parse_timestamp,
+    },
+};
+use ahash::AHashMap;
+use diesel::{upsert::excluded, ExpressionMethods};
+
+// Number of batches processed before gap detected
+pub const GAP_DETECTION_BATCH_COUNT: u64 = 50;
+// Number of batches to process before updating processor status
+const PROCESSOR_STATUS_UPDATE_BATCH_COUNT: u64 = 10;
+
+pub struct GapDetector {
+    processor_name: String,
+    db_pool: PgDbPool,
+    starting_version: u64,
+    seen_versions: AHashMap<u64, ProcessedVersions>,
```

Comment:
why is this so complicated? Just use a queue sorted by time and pop items that don't have a gap. Once you can't pop anymore record the latest version.